File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.java
// Last updated: 9/14/2025, 4:59:03 AM
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int [] res = new int[nums.length];

        int prefix = 1;
        for (int i=0; i<nums.length; i++){
            res[i] = prefix;
            prefix *= nums[i];
        }
        int suffix = 1;
        for (int i = nums.length -1; i >= 0; i--){
            res[i] *= suffix;
            suffix *= nums[i];
        }return res;
    }
}

File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.py
# Last updated: 9/14/2025, 4:59:05 AM
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        """#bruteforce
        n = len(nums)
        res = [0]*n

        for i in range(n):
            prod = 1
            for j in range(n):
                if i==j:
                    continue
                prod *= nums[j]
            res[i] = prod
        return res"""

        #optimal:

        res = [1]*len(nums)

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
        
        suffix = 1
        for i in range(len(nums)-1, -1, -1):
            res[i] *= suffix
            suffix *= nums[i]

        return res
        


        